codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java

```diff
@@ -16,136 +16,240 @@
+        server = http2Handler.connection().isServer();
+        streamKey = connection().newKey();
+        initialLocalConnectionWindow = initialSettings.initialWindowSize();
+    }
+
+    Http2ConnectionHandler connectionHandler() {
+        return http2Handler;
     }
 
     /**
-     * Construct a new handler.
+     * Creates a new outbound/local stream.
+     *
+     * <p>This method may only be called after the handler has been added to a {@link io.netty.channel.ChannelPipeline}.
      *
-     * @param server {@code true} this is a server
+     * <p>This method is thread-safe.
      */
-    public Http2FrameCodec(boolean server, Http2FrameLogger frameLogger) {
-        this(server, new DefaultHttp2FrameWriter(), frameLogger);
-    }
+    Http2Stream2 newStream() {
+        ChannelHandlerContext ctx0 = ctx;
+        if (ctx0 == null) {
+            throw new IllegalStateException("Channel handler not added to a channel pipeline.");
+        }
 
-    // Visible for testing
-    Http2FrameCodec(boolean server, Http2FrameWriter frameWriter, Http2FrameLogger frameLogger) {
-        Http2Connection connection = new DefaultHttp2Connection(server);
-        frameWriter = new Http2OutboundFrameLogger(frameWriter, frameLogger);
-        Http2ConnectionEncoder encoder = new DefaultHttp2ConnectionEncoder(connection, frameWriter);
-        Http2FrameReader frameReader = new DefaultHttp2FrameReader();
-        Http2FrameReader reader = new Http2InboundFrameLogger(frameReader, frameLogger);
-        Http2ConnectionDecoder decoder = new DefaultHttp2ConnectionDecoder(connection, encoder, reader);
-        decoder.frameListener(new FrameListener());
-        http2Handler = new InternalHttp2ConnectionHandler(decoder, encoder, new Http2Settings());
-        http2Handler.connection().addListener(new ConnectionListener());
-        this.server = server;
+        return new Http2Stream2Impl(ctx0.channel());
     }
 
-    Http2ConnectionHandler connectionHandler() {
-        return http2Handler;
+    /**
+     * Iterates over all active HTTP/2 streams.
+     *
+     * <p>This method must not be called outside of the event loop.
+     */
+    void forEachActiveStream(final Http2Stream2Visitor streamVisitor) throws Http2Exception {
+        assert ctx.channel().eventLoop().inEventLoop();
+
+        connection().forEachActiveStream(new Http2StreamVisitor() {
+            @Override
+            public boolean visit(Http2Stream stream) {
+                Http2Stream2 stream2 = stream.getProperty(streamKey);
+                if (stream2 == null) {
+                    /**
+                     * This code is expected to almost never execute. However, in rare cases it's possible that a
+                     * stream is active without a {@link Http2Stream2} object attached, as it's set in a listener of
+                     * the HEADERS frame write.
+                     */
+                    // TODO(buchgr): Remove once Http2Stream2 and Http2Stream are merged.
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)